core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala

```diff
@@ -52,8 +53,20 @@ private[spark] sealed trait MapStatus {
    * partitionId of the task or taskContext.taskAttemptId is used.
    */
   def mapId: Long
-}
 
+  protected def loadLocation(in: ObjectInput): Location = {
+    val conf = SparkEnv.get.conf
+    conf.get(config.SHUFFLE_LOCATION_PLUGIN_CLASS).map { locClass =>
+      val loc = Utils.loadExtensions(
+        classOf[Location],
+        Seq(locClass),
+        conf
+      ).head
+      loc.readExternal(in)
+      loc
```

Comment:
Do we want to explore a `LocationFactory` in addition to `Location` - which is used here ?
`loadExtensions` is not cheap, and doing this for every instance creation is going to be very expensive.

This will also allow for implementations to cache `Location` - else we will have GC issues at driver.

Reply:
I may not understand well about the usage of `LocationFactory` here. But, the problem is, we need a fresh `Location` instance here every time and then make it to a specific one by `readExternal(...)`. So, I'm not sure what do we want to cache here? Or, you mean we can cache the reflected `class`?

Reply:
Both.
`SparkEnv.get.getShuffleLocationFactory.deserialize(in)` (or some such) should make it a method dispatch - with ability for LocationFactory to cache the (immutable) Location object.

Reply:
By "fresh Location instance", I mean, we always need to create the instance using 0-arg constructor first (this's the "fresh" one) and then make it to a specific one by "readExternal".

Follow your idea, I think we cache two things in `LocationFactory`:

1) cache the loaded class of the custom location so we don't need to load it every time

2) cache the location instances so we don't keep the references of multiple instances for the same location

If so, I think we can do it by ourselves instead of exposing `LocationFactory` to users.

Does it sound good to you?

Reply:
I made a prototype at 66977e0. Does it look good to you? (If yes, I'd like to optimize it further to unify the `locationCache` and `blockManagerIdCache` if possible and reduce some duplicate code...)